ompi/mca/coll/acoll/coll_acoll_allgather.c

```diff
@@ -0,0 +1,484 @@
+/* -*- Mode: C; indent-tabs-mode:nil -*- */
+/*
+ * Copyright (C) 2024, Advanced Micro Devices, Inc. All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "ompi_config.h"
+
+
+#include "mpi.h"
+#include "ompi/constants.h"
+#include "ompi/datatype/ompi_datatype.h"
+#include "ompi/mca/coll/base/coll_base_functions.h"
+#include "ompi/mca/coll/base/coll_base_util.h"
+#include "ompi/mca/coll/base/coll_tags.h"
+#include "ompi/mca/coll/coll.h"
+#include "ompi/mca/pml/pml.h"
+#include "opal/util/bit_ops.h"
+#include "coll_acoll.h"
+#include "coll_acoll_utils.h"
+
+/*
+ * coll_allgather_decision_fixed
+ *
+ * Function:    Choose optimal allgather algorithm
+ *
+ * Description: Based on no. of processes and message size, chooses whether
+ *              or not to use subgroups. If subgroup based algorithm is not,
+ *              chosen, further decides if [ring|lin] allgather is to be used.
+ *
+ */
+static inline void coll_allgather_decision_fixed(int size, size_t total_dsize, int sg_size,
+                                                 int *use_ring, int *use_lin)
+{
+    *use_ring = 0;
+    *use_lin = 0;
+    if (size <= (sg_size << 1)) {
+        if (total_dsize >= 1048576) {
+            *use_lin = 1;
+        }
+    } else if (size <= (sg_size << 2)) {
+        if ((total_dsize >= 4096) && (total_dsize < 32768)) {
+            *use_ring = 1;
+        } else if (total_dsize >= 1048576) {
+            *use_lin = 1;
+        }
+    } else if (size <= (sg_size << 3)) {
+        if ((total_dsize >= 4096) && (total_dsize < 32768)) {
+            *use_ring = 1;
+        }
+    } else {
+        if (total_dsize >= 4096) {
+            *use_ring = 1;
+        }
+    }
+}
+
+/*
+ * rd_allgather_sub
+ *
+ * Function:    Uses recursive doubling based allgather for the group.
```

Comment:
have you compared the performance of other methods, besides recursive doubling?

Reply:
Yes, acoll/allgather chooses among recursive doubling, ring and linear based on process count and message sizes.